src/mono/mono/mini/mini.c

```diff
@@ -2840,6 +2849,44 @@ insert_safepoints (MonoCompile *cfg)
 
 }
 
+static void
+insert_samplepoint (MonoCompile *cfg, MonoBasicBlock *bblock)
+{
+	if (cfg->verbose_level > 1)
+		printf ("ADDING SAMPLE POINT TO BB %d\n", bblock->block_num);
+
+	MonoBasicBlock *prev_cbb = cfg->cbb;
+	cfg->cbb = bblock;
+	mini_profiler_emit_samplepoint (cfg);
```

Comment:
This doesn't seem correct. We are inserting code at the end of the basic block. Assuming that it is common for a basic block to end with a branch opcode, this would make the sample point not always hit. For reference, `insert_safepoint` adds the instruction at the beginning of the basic block.

Reply:
@BrzVlad  Could you please suggest/commit a fix, I was struggling with this. Many thanks!

Reply:
A simple solution that I suspect could work would be to create a new tmp_bblock here instead (NEW_BBLOCK) and set it as the cbb. Then you do the emit as before into `tmp_bblock` and once you are done you just move all instructions together from the temporary basic block to the one that you were planning to emit. This would mean that you just update the last_ins of `tmp_bblock` to point to `bblock->code`  and vice versa, and then set `bblock->code` to `tmp_bblock->code`. Hopefully everything would work just like that.